### PR TITLE
feat: sync stars and contributors

### PR DESCRIPTION
**Pull request by TRebirthC**
1. Decouple the logic for storing github history from the projectCrontributors.js.
2. Implement the logic for crawling GitHub star data.
3. Crawl GitHub contributors and stars information on the first day of each month.
4. Fix a bug that the returned contributor list is incorrect when using GitHub API.

### CODE REVIEW

File: packages/integration/controllers/projectContributors.js [hunk 1]
```diff
@@ -171,12 +125,17 @@ async function getContributors(repoName, page = 1) {
       headers: header,
     },
   );
-
   // avoid situations where the project is empty
-  return request.length > 0 ? await request.json() : [];
+  try {
+    const content = await request.json();
+    return content;
```
Comment (Jiannan-dev) [inline packages/integration/controllers/projectContributors.js]: 直接return ， 无须声明content

File: packages/integration/controllers/projectHistory.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  const sumOfProject = projectList.length;
+  logger.info(`The Number of Project : ${sumOfProject}`);
+  let count = 1;
+  for (const project of projectList) {
+    logger.info('**Current Progress**: ', `${count}/${sumOfProject}`);
+    count += 1;
+    // 2. get project information
+    let [contributors, stars] = await getProjectInformation(project.htmlUrl);
+    // check contributors
+    if (contributors == '' || contributors == undefined) {
```
Comment (Jiannan-dev) [inline packages/integration/controllers/projectHistory.js]: 改为falsy判断 if(!contributors)

File: packages/integration/controllers/projectHistory.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      contributors = (await getAlllContributors(project.fullName)).length;
+      logger.info(`GitHub API : contributors of ${project.htmlUrl} is ${contributors}`);
+    }
+    if (contributors == '' || contributors == undefined) {
```
Comment (Jiannan-dev) [inline packages/integration/controllers/projectHistory.js]: 改为falsy判断 if(!contributors)

File: packages/integration/controllers/projectHistory.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      continue;
+    }
+    // check stars
+    if (stars == '' || stars == undefined) {
```
Comment (Jiannan-dev) [inline packages/integration/controllers/projectHistory.js]: 改为falsy判断

File: packages/integration/controllers/projectHistory.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      stars = await getStars(project.fullName);
+      logger.info(`GitHub API : stars of ${project.htmlUrl} is ${stars}`);
+    }
+    if (stars == '' || stars == undefined) {
```
Comment (Jiannan-dev) [inline packages/integration/controllers/projectHistory.js]: 改为falsy判断

**Review by Jiannan-dev — CHANGES_REQUESTED** [review]

File: packages/integration/controllers/projectHistory.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    count += 1;
+    // 2. get project information
+    let [contributors, stars] = await getProjectInformation(project.htmlUrl);
+    // check contributors
+    if (contributors == '' || contributors == undefined) {
+      contributors = (await getAlllContributors(project.fullName)).length;
+      logger.info(`GitHub API : contributors of ${project.htmlUrl} is ${contributors}`);
+    }
+    if (contributors == '' || contributors == undefined) {
+      continue;
```
Comment (Jiannan-dev) [inline packages/integration/controllers/projectHistory.js]: 无contributor时 continue是否会导致获取star的逻辑跳过？

File: packages/integration/controllers/projectHistory.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        } else {
+          realNumber = parseInt(contributorsNumMain);
+        }
+        contributors = realNumber.toString();
+
+        logger.info(`web crawler: contributors of ${url} is ${contributors}`);
+      }
+      // get star
+      const spanStar = $('#repo-stars-counter-star');
+      if (spanStar === null) {
```
Comment (Jiannan-dev) [inline packages/integration/controllers/projectHistory.js]: fasly

File: packages/integration/controllers/projectHistory.js [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        logger.info(`web crawler: ${url} does not provide stars...`);
+      } else {
+        const starValueOrigin = spanStar.attr('title');
+        stars = starValueOrigin.replace(/,/g, '');
```
Comment (Jiannan-dev) [inline packages/integration/controllers/projectHistory.js:126]: 是否有对 没有star ， star过多(以 n k显示)情况的兼容？

**Review by Jiannan-dev — CHANGES_REQUESTED** [review]

File: packages/integration/controllers/projectHistory.js (hunk 8 above)
Comment (TRebirthC) [reply packages/integration/controllers/projectHistory.js:126]: 通过该手段返回的star数是不带k的数值，即时GitHub首页显示的是多少k，取到的数值也可以精确到个位。

File: packages/integration/controllers/projectContributors.js (hunk 1 above)
Comment (TRebirthC) [reply packages/integration/controllers/projectContributors.js]: 已在 [commit] [fix: minor syntax and logic changes](https://github.com/originjs/oss-evaluation-service/pull/677/commits/051299abe9f0bad28e7b0e9329d40b90f1daf28c) 中进行修改

File: packages/integration/controllers/projectHistory.js (hunk 2 above)
Comment (TRebirthC) [reply packages/integration/controllers/projectHistory.js]: 已在 [commit] [fix: minor syntax and logic changes](https://github.com/originjs/oss-evaluation-service/pull/677/commits/051299abe9f0bad28e7b0e9329d40b90f1daf28c) 中进行修改

File: packages/integration/controllers/projectHistory.js (hunk 3 above)
Comment (TRebirthC) [reply packages/integration/controllers/projectHistory.js]: 已在 [commit] [fix: minor syntax and logic changes](https://github.com/originjs/oss-evaluation-service/pull/677/commits/051299abe9f0bad28e7b0e9329d40b90f1daf28c) 中进行修改

File: packages/integration/controllers/projectHistory.js (hunk 4 above)
Comment (TRebirthC) [reply packages/integration/controllers/projectHistory.js]: 已在 [commit] [fix: minor syntax and logic changes](https://github.com/originjs/oss-evaluation-service/pull/677/commits/051299abe9f0bad28e7b0e9329d40b90f1daf28c) 中进行修改

File: packages/integration/controllers/projectHistory.js (hunk 5 above)
Comment (TRebirthC) [reply packages/integration/controllers/projectHistory.js]: 已在 [commit] [fix: minor syntax and logic changes](https://github.com/originjs/oss-evaluation-service/pull/677/commits/051299abe9f0bad28e7b0e9329d40b90f1daf28c) 中进行修改

File: packages/integration/controllers/projectHistory.js (hunk 6 above)
Comment (TRebirthC) [reply packages/integration/controllers/projectHistory.js]: 会，已在 [commit] [fix: minor syntax and logic changes](https://github.com/originjs/oss-evaluation-service/pull/677/commits/051299abe9f0bad28e7b0e9329d40b90f1daf28c) 中进行修改，将该逻辑下移到获取star之后。

File: packages/integration/controllers/projectHistory.js (hunk 7 above)
Comment (TRebirthC) [reply packages/integration/controllers/projectHistory.js]: 已在 [commit] [fix: minor syntax and logic changes](https://github.com/originjs/oss-evaluation-service/pull/677/commits/051299abe9f0bad28e7b0e9329d40b90f1daf28c) 中进行修改

**Review by Jiannan-dev — APPROVED** [review]

**Pull request merged**